File: src/java/magma/result/Err.java
package magma.result;

import java.util.function.Function;

public record Err<T, X>(X error) implements Result<T, X> {
    @Override
    public <R> Result<R, X> flatMapValue(Function<T, Result<R, X>> mapper) {
        return new Err<>(this.error);
    }

    @Override
    public <R> Result<R, X> mapValue(Function<T, R> mapper) {
        return new Err<>(this.error);
    }

    @Override
    public <R> Result<T, R> mapErr(Function<X, R> mapper) {
        return new Err<>(mapper.apply(this.error));
    }

    @Override
    public <R> R match(Function<T, R> onOk, Function<X, R> onErr) {
        return onErr.apply(this.error);
    }
}

File: src/c/magma/result/Err.c
#include <temp.h>
#include <temp.h>
#include <temp.h>
struct Err<T, X> {
	X error;
	struct Err<T, X> new(X error){
		struct Err<T, X> this;
		this.error = error;
		return this;
	}
	X> flatMapValue(void* _this_){
		struct Err<T, X> this = *(struct Err<T, X>*) this;
		temp = temp;
	}
	X> mapValue(void* _this_){
		struct Err<T, X> this = *(struct Err<T, X>*) this;
		temp = temp;
	}
	R> mapErr(void* _this_){
		struct Err<T, X> this = *(struct Err<T, X>*) this;
		temp = temp;
	}
	R match(void* _this_){
		struct Err<T, X> this = *(struct Err<T, X>*) this;
		return onErr.apply(this.error);
	}
	X> and(void* _this_){
		struct Err<T, X> this = *(struct Err<T, X>*) this;
		temp = temp;
	}
	R>> or(void* _this_){
		struct Err<T, X> this = *(struct Err<T, X>*) this;
		return other.get().mapErr(otherErr -> new Tuple<>(this.error, otherErr));
	}
};